kernel/darwin/iokit/driver.cpp

```diff
@@ -0,0 +1,153 @@
+#include <sys/proc.h>
+#include <sys/vnode.h>
+
+#include "kernel/darwin/iokit/public.h"
+#include "kernel/darwin/iokit/logging.h"
+
+// This required macro defines the class's constructors, destructors,
+// and several other methods I/O Kit requires.
+OSDefineMetaClassAndStructors(com_facebook_iokit_OSQuery, IOService)
+
+// Define the driver's superclass.
+#define super IOService
+
+// Paths to monitor. File modifications in any subdirectory will be logged.
+const char *const monitored_paths[] = { "/Users", "/Library", "/Volume",
+                                        "/System" };
+
+// Number of active listener instances.
+SInt32 listener_instances = 0;
+
+static void SendMessageToDaemon(
+    IOSharedDataQueue *data_queue, kauth_cred_t credential,
+    kauth_action_t action, uintptr_t arg0, uintptr_t arg1) {
+  kernel_osquery_message_t message = {0};
+  clock_get_calendar_microtime(
+      &(message.time_secs), &(message.time_microsecs));
+  message.action = action;
+  message.pid = proc_selfpid();
+  message.ppid = proc_selfppid();
+  message.uid = kauth_cred_getuid(credential);
+  message.ruid = kauth_cred_getruid(credential);
+  proc_selfname(message.proc_name, MAXCOMLEN + 1);
+  if (arg0)
+    strlcpy(message.from_path, reinterpret_cast<const char*>(arg0),
+            MAX_PATH_LEN + 1);
+  if (arg1)
+    strlcpy(message.to_path, reinterpret_cast<const char*>(arg1),
+            MAX_PATH_LEN + 1);
+  if (data_queue)
+    data_queue->enqueue(&message, sizeof(message));
+}
+
+// The kernel ignores the return value of a file operation scope listener.
+// However, it is recommended that the function always return
+// KAUTH_RESULT_DEFER.
+static int FileOpScopeListener(
+    kauth_cred_t credential, void *data_queue, kauth_action_t action,
+    uintptr_t arg0, uintptr_t arg1, uintptr_t arg2, uintptr_t arg3) {
+  OSIncrementAtomic(&listener_instances);
+  bool valid_event = true;
+  // Ignore noisy events.
+  if (action == KAUTH_FILEOP_OPEN ||
+      (action == KAUTH_FILEOP_CLOSE && !(arg2 & KAUTH_FILEOP_CLOSE_MODIFIED)))
+    valid_event = false;
+  if (valid_event && action != KAUTH_FILEOP_EXEC) {
+    valid_event = false;
+    // Check whether the file path matches one of the monitored paths.
+    for (const auto& path : monitored_paths) {
+      if (strprefix(reinterpret_cast<const char *>(arg1), path)) {
+        valid_event = true;
+        break;
+      }
+    }
+  }
+  if (valid_event)
+    SendMessageToDaemon(
+        reinterpret_cast<IOSharedDataQueue *>(data_queue), credential, action,
+        arg0, arg1);
+  OSDecrementAtomic(&listener_instances);
+  return KAUTH_RESULT_DEFER;
+}
+
+bool OSQueryDriverClassName::start(IOService *provider) {
+  bool result = super::start(provider);
+  LOGD("Starting.");
+  // Create and initialize an IODataQueue object.
+  data_queue_ = IOSharedDataQueue::withCapacity(
+      kQueueEventCapacity
+      * (sizeof(kernel_osquery_message_t) + DATA_QUEUE_ENTRY_HEADER_SIZE));
+  if (!data_queue_) {
```

Comment:
I'm a big fan on explicit comparisons to `nullptr` when appropriate. But definitely up to you.